packages/vats/src/proposals/network-proposal.js

```diff
@@ -1,20 +1,85 @@
-import { E, Far } from '@endo/far';
+import { E } from '@endo/far';
 import { BridgeId as BRIDGE_ID } from '@agoric/internal';
 import {
-  makeLoopbackProtocolHandler,
-  makeEchoConnectionHandler,
-  makeNonceMaker,
+  prepareEchoConnectionHandler,
+  prepareNonceMaker,
+  prepareLoopbackProtocolHandler,
 } from '@agoric/network';
 
+// NOTE: Heap-based whenable resolution is used for this module because the
+// bootstrap vat can't yet be upgraded.
+import { makeDurableZone } from '@agoric/zone/durable.js';
+import { prepareWhenableModule } from '@agoric/whenable';
+
 const NUM_IBC_PORTS_PER_CLIENT = 3;
 const INTERCHAIN_ACCOUNT_CONTROLLER_PORT_PREFIX = 'icacontroller-';
 
 /**
+ * @param {import('@agoric/base-zone').Zone} zone
+ * @param {ReturnType<prepareEchoConnectionHandler>} makeEchoConnectionHandler
+ */
+const prepareListenHandler = (zone, makeEchoConnectionHandler) => {
+  const makeListenHandler = zone.exoClass(
+    'listener',
+    undefined,
+    () => {
+      return {
+        handler: makeEchoConnectionHandler(),
+      };
+    },
+    {
+      async onAccept(_port, _localAddr, _remoteAddr, _listenHandler) {
+        return harden(this.state.handler);
+      },
+      async onListen(port, _listenHandler) {
+        console.debug(`listening on echo port: ${port}`);
+      },
+    },
+  );
+
+  return makeListenHandler;
+};
+
+/**
+ * @param {import('@agoric/base-zone').Zone} zone
+ * @param dibcBridgeManager
+ */
+const prepareCallbacks = (zone, dibcBridgeManager) => {
+  return zone.exoClass('callbacks', undefined, () => ({}), {
+    downcall(method, obj) {
+      return E(dibcBridgeManager).toBridge({
+        ...obj,
+        type: 'IBC_METHOD',
+        method,
+      });
+    },
+  });
+};
+
+/**
+ * @param {import('@agoric/base-zone').Zone} zone
  * @param {SoloVats | NetVats} vats
  * @param {ERef<import('../types.js').ScopedBridgeManager>} [dibcBridgeManager]
  */
-export const registerNetworkProtocols = async (vats, dibcBridgeManager) => {
+export const registerNetworkProtocols = async (
+  zone,
+  vats,
+  dibcBridgeManager,
+) => {
   const ps = [];
+  const makeNonceMaker = prepareNonceMaker(zone);
+  const makeLoopbackProtocolHandler = prepareLoopbackProtocolHandler(
```

Comment:
This puts the loopback protocol handler in the bootstrap vat, for which upgrade is currently infeasible (#8849). Is that by design? Did you consider putting it in the network vat?

Reply:
The result of our pairing session: cfc162e5e thru d49e2d751
I added a bit after that: 654a4857f
branch: `dc-ibc-net-review`